builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,598 @@
+			resource.WithName(rmResourcePromotedJobName),
+			resource.WithState(&Resource_Job{}),
+			resource.WithCreate(r.resourceJobCreate),
+			resource.WithDestroy(r.resourceJobDestroy),
+			resource.WithStatus(r.resourceJobStatus),
+			resource.WithPlatform("nomad-jobspec"),
+			resource.WithCategoryDisplayHint(sdk.ResourceCategoryDisplayHint_INSTANCE_MANAGER),
+		)),
+	)
+}
+
+// getNomadClient provides
+// the client connection used by resources to interact with Nomad.
+func (r *Releaser) getNomadClient() (*nomadClient, error) {
+	// Get our client
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		return nil, err
+	}
+	return &nomadClient{
+		NomadClient: client,
+	}, nil
+}
+
+func (r *Releaser) resourceJobCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	target *Deployment,
+	result *Release,
+	state *Resource_Job,
+	client *nomadClient,
+	st terminal.Status,
+) error {
+	// Set up clients
+	jobClient := client.NomadClient.Jobs()
+	deploymentClient := client.NomadClient.Deployments()
+
+	st.Update("Getting job...")
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	}
+
+	if len(jobs) > 0 {
+		if target.Name != jobs[0].ID {
+			st.Step(terminal.StatusError, fmt.Sprintf("Job could not be found, did you mean to promote %q?", jobs[0].ID))
+			return nil
+		}
+	} else {
+		st.Step(terminal.StatusError, "Job not found.")
+		return nil
+	}
+
+	q := &api.QueryOptions{
+		Namespace: jobs[0].JobSummary.Namespace,
+	}
+
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job.")
+		return nil
+	} else if deploy.JobVersion != *job.Version {
+		st.Update("Job version does not match deployment's job version.")
+		return nil
+	} else if deploy.Status != "running" {
+		st.Update("Deployment for job is no longer running.")
+		return nil
+	}
+
+	canaryDeployment := false
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries > 0 {
+				canaryDeployment = true
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries > 0 {
+			canaryDeployment = true
+			// If no groups to promote are specified in the config, promote all groups
+			//   that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
+	}
+	if !canaryDeployment {
```

Comment:
Is this technically a release failure? If there's no canaries defined in the jobspec file, should we abort the release? Right now we say the release was successful in this case.

```
... deploy output ....
✓ Job "web" is reporting ready!

» Releasing example-nodejs...
⚠️ Nomad canary allocations not detected in job task groups.
✓ Release successfully rolled out!

✓ Finished building report for Nomad platform
✓ Finished building report for Nomad job resource
```

Reply:
I noticed this happens on the very _first_ run. No canary allocations are detected in the task group, and the second run, they exist 🤔

Reply:
Yep, canaries are introduced in the `update` stanza, and since the very first deployment of a job isn't considered an "update", all allocations are deployed right away. I think it'd make sense to add something to check if it's `jobVersion` 0, and then not report an error here!

Reply:
That sounds good to me!